# Домашнее задание по уроку «Случайные события. Случайные величины»

Смоделировать игру против лотерейного автомата типа “777”. Игрок платит 1 руб., после чего выпадает случайное целое число, равномерно распределенное от 0 до 999. При некоторых значениях числа игрок получает выигрыш (см. справа)

Выгодна ли игра игроку?

Сколько в среднем приобретает или теряет игрок за одну игру?

777: 200 руб.  
999: 100 руб.  
555: 50 руб.  
333: 15 руб.  
111: 10 руб.  
x77: 5 руб.  
xx7: 3 руб.  
x00: 2 руб.  
xx0: 1 руб.  

x – любая цифра

Дополнительное задание повышенной сложности. Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования

In [1]:
import pandas as pd
import numpy as np
from decimal import Decimal

In [2]:
def get_money_prize_type_777() -> int:
    """
    Функция для определения суммы выигрыша в лоторейном автомате типа 777.
    
    Результат: сумма выигрыша, тип int 
    """
    
    number = np.random.randint(0, 1000)
        
    if number == 777:
        return 200 
    elif number == 999:
        return 100
    elif number == 555:
        return 50
    elif number == 333:
        return 15  
    elif number == 111:
        return 10
    elif number%100 == 77:
        return 5
    elif number%10 == 7:
        return 3
    elif number%100 == 0:
        return 2
    elif number%10 == 0:
        return 1
    else:
        return 0  

In [3]:
play_count = 10000000
win_moneys = 0
lose_moneys = 0

for _ in range(0, play_count):
    lose_moneys += 1
    win_moneys += get_money_prize_type_777()

In [4]:
# Средний размер выигрыша за одну игру
win_moneys/play_count

0.8007059

In [5]:
# Средний размер проигрыша за одну игру
lose_moneys/play_count

1.0

# Рассчитаем теоретический размер проигрыша за одну игру.

Для начала нам нужно посчитать число выигрышных комбинаций и их вероятность.

777 - 1 комбинация из 1000 (вероятность попадания 0.001)

999 - 1 комбинация из 1000 (вероятность попадания 0.001)

555 - 1 комбинация из 1000 (вероятность попадания 0.001)

333 - 1 комбинация из 1000 (вероятность попадания 0.001)

333 - 1 комбинация из 1000 (вероятность попадания 0.001)

x77 - 9 комбинаций из 1000 (вычитаем из общего количества комбинаций 10 одну комбинацию в виде 777, вероятность попадания 0,009)

xx7 - 90 комбинации из 1000 ($\overline{A}_{10}^2 - 10 = 10^2 - 10 = 90$, используем формулу размещения с повторением и вычитаем 10 комбинаций типа x77) (вероятность попадания 0.09)

x00 - 10 комбинаций из 1000 (вероятность попадания 0.01)

xx0 - 90 комбинаций из 1000 ($\overline{A}_{10}^2 - 10 = 10^2 - 10 = 90$, используем формулу размещения с повторением и вычитаем 10 комбинаций типа x00, вероятность попадания 0.09)

По формуле математического ожидания случайной величины получаем рассчетное среднее значение выигрыша:

In [6]:
(Decimal('0.001')*200 + 
 Decimal('0.001')*100 + 
 Decimal('0.001')*50 + 
 Decimal('0.001')*15 + 
 Decimal('0.001')*10 +
 Decimal('0.009')*5 + 
 Decimal('0.09')*3 +
 Decimal('0.01')*2 + 
 Decimal('0.09')*1
)

Decimal('0.800')

Средний выигрыш за одну игру по методу Монте-Карло - 0.8007059

Теоретический средний выигрыш - 0.800

Теоретический размер проигрыша будет 1, так как в каждой игре человек тратит постоянную сумму.